File: F. Amuse-Bouche/HeathrowToLondon.hs
module Main where

-- import Data.Function (on)
import Data.List (intersperse, sort)
-- trace is used to display the input and output of functions.
import Debug.Trace

-- By convention, modules should have a main function,
-- which is how the code in the module starts on its own.
-- Alternatively, one can call optimalPaths directly from the prompt.
main = do putStrLn . show . optimalPaths $ [50, 10, 30, 5, 90, 20, 40, 2, 25, 10, 8]


-- The A side, the B side, or a Crossover.
data Label = A | B | C deriving (Eq, Show)

-- A Step is a transition from one intersection to another.
-- It is on the A or B side, or it's a crossover. In all
-- cases it has an associated distance.
data Step = Step Label Int

-- Define how show runs on a Step. Makes Step an instance of the Show typeclass.
instance Show Step where
  show (Step lbl dist) = show lbl ++ "/" ++ show dist

--   A Path is a sequence of Steps.
data Path = Path [Step]

-- The distance for a Path is the sum of the Step distances.
dist :: Path -> Int
dist (Path steps) = sum $ map (\(Step _ d) -> d) steps
-- The following is also correct.
-- dist (Path steps) = foldl (\sum (Step _ d) -> sum+d) 0 steps

-- measureOfMerit is a metric for how good a path is. The
-- shorter the distance the better. If two Paths have the
-- same distance, fewer steps is better.
measureOfMerit :: Path -> (Int, Int)
measureOfMerit p@(Path steps) = (dist p, length steps)

-- Two Paths are equal if their measures of merit are equal.
instance Eq Path where
      -- ((+) `on` f) x y = f x + f y (not using `on` from Data.Function)
  (==) = (==) `on` measureOfMerit

-- This says that Paths can be compared.
-- Compare paths by comparing their measures of merit.
instance Ord Path where
  compare = compare `on` measureOfMerit

-- Instead of importing it from Data.Function, this is how `on` works.
-- E.g., compare = compare `on` measureOfMerit means that when comparing
-- two elements apply measureOfMerit to each and then compare those results.
-- In this case f1 is compare; f2 is measureOfMerit
on :: (b -> b -> c) -> (a -> b) -> (a -> a -> c)
f1 `on` f2 = \x y -> f1 (f2 x) (f2 y)


-- The start of a path is the label (A | B | C) of the first step.
start :: Path -> Label
start (Path (Step label _ : _)) = label

-- If the label of the last step is A or B, that's the
-- value for end. If the last step is C, the path ends
-- at the opposite of the previous step.
end :: Path -> Label
end (Path steps) =
  case reverse steps of
  Step C _ : rest     -> case rest of Step A _ : _ -> B
                                      Step B _ : _ -> A
  Step lastLbl _ : _  -> lastLbl

-- Define show on Path.
-- Notice that each Path string starts on a new line.
instance Show Path where
  show p@(Path steps) =
    "\n" ++ show (start p) ++ "->" ++ show (end p)
    ++ ". Dist: " ++ show (dist p) ++ ". "
    ++ (concat . intersperse ", " . map show) steps

-- Can two paths be joined, i.e., does the first one end
-- where the second one starts? Use a new infix operator,
-- which can be defined on the fly.
(>?<) :: Path -> Path -> Bool
p1 >?< p2 = end p1 == start p2

-- Extend one path by another by appending their Steps.
-- Should not be used unless the Paths can be joined.
(+>) :: Path -> Path -> Path
-- A guard may be used to check the >?< constraint.
p1@(Path steps1) +> p2@(Path steps2)
  | p1 >?< p2 = Path (steps1 ++ steps2)
  | otherwise = error ("Attempt to join " ++ show (end p1)
                       ++ " to " ++ show (start p2) ++ " in "
                       ++ show p1 ++ " +> " ++ show p2 ++ "\n")


-- A QuadPaths object always has exactly 4 paths connecting from
-- one cross link to another. It has the four optimal paths for:
-- A-to-A, A-to-B, B-to-A, B-to-B.
data QuadPaths = QuadPaths [Path] deriving Show

-- Join two Quadpaths by joining the paths that fit and selecting
-- the best path for each start-end combination. This is another new infix operator.
-- E.g. A-to-B can be either A-to-A +> A-to-B or A-to-B +> B-to-B.
(++>) :: QuadPaths -> QuadPaths -> QuadPaths
(QuadPaths paths1) ++> (QuadPaths paths2) =
  let result = trace ("\njoin (input):\n" ++ show (QuadPaths paths1) ++ show (QuadPaths paths2))
                     join (QuadPaths paths1) (QuadPaths paths2)
  in trace ("\njoin (output):\n" ++ show result ++ "\n") result

join (QuadPaths paths1) (QuadPaths paths2) =
   QuadPaths [ bestPath s e | s <- [A, B], e <- [A, B]] where
      bestPath :: Label -> Label -> Path
      bestPath s e = minimum [p1 +> p2 | p1 <- paths1, s == start p1,
                                         p2 <- paths2, e == end p2,
                                         p1 >?< p2]

  -- An alternative formuation:
  -- QuadPaths (map bestWithStartEnd [(A, A), (A, B), (B, A), (B, B)])
  -- where bestWithStartEnd :: (Label, Label) -> Path
  --       bestWithStartEnd (s, e) =
  --         minimum [p1 +> p2 | p1 <- paths1, s == start p1,
  --                             p2 <- paths2, e == end p2,
  --                             p1 >?< p2]

-- Input to the overall problem is given as a list of Ints. The
-- list should be understood as sequences of three Steps.
-- Each 3-Step sequence is (a) the distance along the A road,
-- (b) the distance along the B road, and (c) the distance of the
-- crossover that connects the ends of the two roads.

-- The problem is to find all 4 shortest paths from Heathrow
-- to London. When sorted the head is the minimum.

-- The program first uses qpList to convert the initial int list into
-- a sequence of QuadPaths. Then it uses combinePaths to combine
-- the list of QuadPaths into a single QuadPath. Then it sorts the
-- paths in that QuadPath.

optimalPaths :: [Int] -> [Path]
optimalPaths input =
  let result = optimalPaths' $ trace ("\noptimalPaths (input):\n" ++ show input) input
  in trace ("\noptimalPaths (output):\n" ++ show result ++ "\n") result

optimalPaths' = sort . (\(QuadPaths paths) -> paths) . combinePaths . qpList
  -- where
  -- combinePaths glues a list of QuadPaths objects together.

  -- Gluing two QuadPaths together produces a QuadPaths object
  -- with four optimum paths (A-to-A, A-to-B, B-to-A, B-to-B)
  -- over the distance of the two objects together. (See ++>.)

  -- Any of the following versions of combinePaths works.

  -- combinePaths :: [QuadPaths] -> QuadPaths
  -- combinePaths = foldl1 (++>)

  -- combinePaths :: [QuadPaths] -> QuadPaths
  -- combinePaths = foldr1 (++>)

  -- The following version is longer, but it lends itself to
  -- parallel processing. It also illustrates that the sequence
  -- in which the QuadPaths objects are glued together doesn't
  -- matter.
  --
  -- Split the QuadPaths list into two lists; combine the
  -- QuadPaths in each list; combine the two results.

combinePaths :: [QuadPaths] -> QuadPaths
combinePaths qps =
  let result = combinePaths' $ trace ("\ncombinePaths (input):\n" ++ show qps) qps
  in trace ("\ncombinePaths (output):\n" ++ show result ++ "\n") result

combinePaths' (qp : qps) = foldl (++>) qp qps
-- combinePaths' [qp] = qp
-- combinePaths' [qp1, qp2] = qp1 ++> qp2
-- combinePaths' qps = combinePaths leftQPs ++> combinePaths rightQPs
--       where (leftQPs, rightQPs) = splitAt (length qps `div` 2) qps


-- Read the inputs 3 at a time and make a QuadPaths of each group.
-- Each QuadPaths object is the 4 paths for a segment.
qpList :: [Int] -> [QuadPaths]
qpList list =
  let result = qpList' $ trace ("\nqpList (input):\n" ++ show list) list
  in trace ("\nqpList (output):\n" ++ show result ++ "\n") result

qpList' (aDist : bDist : cDist : rest) =
  QuadPaths [ Path [Step A aDist],               -- A to A
              Path [Step A aDist, Step C cDist], -- A to B
              Path [Step B bDist, Step C cDist], -- B to A
              Path [Step B bDist]                -- B to B
            ] : qpList' rest
qpList' [] = []
-- What to do if the number of input elements is not a multiple of 3.
-- These are arbitrary decisions.
qpList' [aDist, bDist] = qpList' [aDist, bDist, 0]
qpList' [xDist]        = qpList' [xDist, xDist, 0]



{-
   Test using the example in the book.
   > optimalPaths [50, 10, 30, 5, 90, 20, 40, 2, 25, 10, 8]
   [
   B->B. Dist: 75. B/10, C/30, A/5, C/20, B/2, B/8,
   B->A. Dist: 75. B/10, C/30, A/5, C/20, B/2, B/8, C/0,
   A->B. Dist: 85. A/50, A/5, C/20, B/2, B/8,
   A->A. Dist: 85. A/50, A/5, C/20, B/2, B/8, C/0]


   Why does changing the second crossover from 15 to 14 in
   the following examples cause the shortest path to take
   the first crossover?

   > optimalPaths [10, 50, 30, 50, 5, 15, 2, 40, 0, 10, 5]
   [
   A->B. Dist: 67. A/10, A/50, A/2, C/0, B/5,
   A->A. Dist: 67. A/10, A/50, A/2, C/0, B/5, C/0,
   B->B. Dist: 77. B/50, B/5, C/15, A/2, C/0, B/5,
   B->A. Dist: 77. B/50, B/5, C/15, A/2, C/0, B/5, C/0]

   > optimalPaths [10, 50, 30, 50, 5, 14, 2, 40, 0, 10, 5]
   [
   A->B. Dist: 66. A/10, C/30, B/5, C/14, A/2, C/0, B/5,
   A->A. Dist: 66. A/10, C/30, B/5, C/14, A/2, C/0, B/5, C/0,
   B->B. Dist: 76. B/50, B/5, C/14, A/2, C/0, B/5,
   B->A. Dist: 76. B/50, B/5, C/14, A/2, C/0, B/5, C/0]

-}

File: F. Amuse-Bouche/HeathrowToLondon.py
from functools import reduce, wraps
from typing import Any, Callable, List, Tuple, TypeVar

# Road derives from type. Type is a metaclass, so Road is a metaclass. A metaclass
# is basically the class of a class, defining how that class behaves.
class Road(type):
    """
    There are three Roads (i.e., Road Types) in the problem: A, B, and C.

    Each Road Type is both an instance of the type Road as well as a class with its own (Segment) instances.

    The Road type itself, i.e., this class, is considered a metatype (and is a subtype of type) because it
    has types (A, B, and C) as instances.

    Roads are divided into Segments. We use the types A, B, and C when we declare those instances,
    i.e., theose Segments. The A-Road and B-Road Segments are each connected to each other.
    The Segments of the C road are not connected to each other.

    The Segment type enables us to talk about all segments as instances of a common type, i.e., the Segment type.
    To do that we make the Segment type a supertype of each of the Road types.

    The Road type also defines __str__ on Segments. This could also be done at the Segment level.
    """
    def __str__(cls):
        return cls.__name__

# Explicitly defines Segment's metaclass to be Road rather than the default type.
# The metaclass works as a template for class creations.
class Segment(metaclass=Road):
    """
    The Segment class is an artifical class used to represent the common features of Segments of the
    three Road types.

    As indicated above, each Road type is an instance of the metatype Road. (Road is a metatype because
    it has types, i.e., classes, as instances.

    Segment has three subtypes: A, B, and C.

    Since each Road type is an instance of the type Road, a Segment is also an instance of the type Road.
    In other words, a Segment is both an instance of Road and a type for Segment instances.

    Each Segment instance has a distance.
    """

    def __init__(self, dist: int):
        self.dist = dist

    def __str__(self):
        """
        This defines the string representation of a segment, i.e., A/5.
        (The format is the same as in the Haskell code.)

        Note: In the following, type(self) is the class itself, i.e., A, B, or C.
        When we call str(type(class)) we access __str__ in Road.
        The result is the same as calling type(self).__name__
        """
        return f'{str(type(self))}/{self.dist}'


class A(Segment):
    """
    This class itself is an instance of the metaclass Road.

    This class also has instances, each of which is a Segment.

    Similarly for the B and C classes.
    """
    def __init__(self, dist: int):
        super().__init__(dist)

class B(Segment):
    """
    A segment of the B road.
    """

    def __init__(self, dist: int):
        super().__init__(dist)

class C(Segment):
    """
    A segment of the C road.
    """

    ...   # This is like pass


class Path:

    # Methods that begin and end with double underscores are defined and used at the Python level.
    def __init__(self, startingRoad: Road, endingRoad: Road, steps: List[Segment]):
        """
        Here we are using Road objects as parameters. Those Road objects are instances of
        the metatype Road.
        """
        self.startingRoad = startingRoad
        self.endingRoad = endingRoad
        self.steps = steps

    def __add__(self, otherPath: 'Path') -> 'Path':
        """
        The method that implements to the (+) operator for this class.
        """
        return Path(self.startingRoad, otherPath.endingRoad, self.steps + otherPath.steps)

    def __str__(self) -> str:
        """
        The method that returns the common printable form of instances for this class.
        It's like toString() and Java and show() in Haskell.
        """
        st = f'{self.startingRoad}->{self.endingRoad}. Dist: {self._dist()}. ' + \
             listToString(self.steps, start='', end='')
        return st

    # Methods that begin with single underscores are taken by convention to be private.
    def _dist(self) -> int:
        return sum( (step.dist for step in self.steps) )

    def _numSteps(self):
        return len(self.steps)

    # This method is taken as the equivalent of public.
    def figureOfMerit(self) -> Tuple[int, int]:
        """
        When comparing two paths, the one with the shorter distance is better.
        If they have the same distance, the one with the fewer steps is better.
        This function returns a value that can be compared to that of other Paths.
        """
        return (self._dist(), self._numSteps())


class QuadPaths:
    """
    The shortest Paths from A and B to A and B in all combinations.

    The class name is plural because each instance contains four Paths.
    """

    def __init__(self, paths: List[Path]):
        self.paths = paths

    def __str__(self) -> str:
        st = listToString(self.paths, start='QuadPaths:\n    ', sep='\n    ', end='')
        return st

# T2 can be any type
T2 = TypeVar('T2')
def trace(func: Callable[..., T2]) -> Callable[..., T2]:
    """
    Print the function signature and return value.
    Adapted from the @debug decorator of Hjelle, Primer on Python Decorators
    (https://realpython.com/primer-on-python-decorators/#debugging-code)
    """
    # https://docs.python.org/2/library/functools.html
    @wraps(func)
    def wrapper_trace(*args: List[Any], **kwargs: List[Any]) -> T2:
        # make list of strings of args
        args_str = [str(a) for a in args]
        # make list of strings of k=v for each entry in kwargs
        kwargs_str = [f'{k}={str(v)}' for (k, v) in kwargs.items()]
        # makes string from list by concatenating the above, using \n to start, end,
        # and delimite each item in the list
        fullArgsStr = listToString(args_str + kwargs_str, start='\n', sep=',\n', end='\n')
        print(f'\nCalling {func.__name__}({fullArgsStr})')
        # value = the return of calling the actual function
        value = func(*args, **kwargs)
        # uses custom listToString if list, otherwise use class defined __str__
        valueStr = str(value) if type(value) is not list else listToString(value, start='', sep=',\n', end='\n')
        print(f'{func.__name__} returned: \n{valueStr}\n')
        return value

    # Prints the trace info from wrapper_trace and returns actual return from func
    return wrapper_trace

def bestPath(startingRoad: Road, endingRoad: Road, qp1: QuadPaths, qp2: QuadPaths) -> Path:
    """
    Find the pair of Paths from qp1 and qp2 such that:
      o the qp1 Path starts at the startingRoad;
      o the qp2 Path starts where the qp1 Path ends;
      o the qp2 Path ends at the endingRoad.
      o the combined Path is the shortest such combination.

    Join those two Paths into a single combined Path and return that combined Path.

    Note: (+) is defined for Path to join two Paths into a new Path.
    See Path.__add__().
    """
    paths = [p1 + p2 for p1 in qp1.paths if p1.startingRoad == startingRoad
                     for p2 in qp2.paths if p1.endingRoad == p2.startingRoad and p2.endingRoad == endingRoad]

    # Python is smart enough to know that Path.figureOfMerit is a method of the Path class
    # and should be applied to a path p as p.figureOfMerit()

    # sorted() returns a new list and takes an iterable, optional key for basing
    # comparison, and optional bool for reverse. sortd[0] returns the first path
    # of the sorted list which is the "best" (shortest/least steps) path
    sortd = sorted(paths, key=Path.figureOfMerit)
    return sortd[0]

# @trace wraps joinQuadPaths in the trace wrapper
@trace
def joinQuadPaths(qp1: QuadPaths, qp2: QuadPaths) -> QuadPaths:
    """
    Joins two adjacent QuadPaths objects into a combined QuadPaths object.
    """
    # list comp applying bestPath to each possible start and end combination on
    # the two adjoining QuadPaths. A QuadPath is constructed from the list
    joinedQuadPaths = QuadPaths([bestPath(s, e, qp1, qp2) for s in [A, B] for e in [A, B]] )
    return joinedQuadPaths

# T1 can be any type
T1 = TypeVar('T1')
def listToString(aList: List[T1], start='[', sep=', ', end=']') -> str:
    """
    A utility that allows one to add start and end strings to a joined string.
    The default is to use '[' and ']' as start and end as a way to print lists.
    """
    return start + sep.join([str(elt) for elt in aList]) + end

#
def optimalPath(allSegs: List[int]) -> Path:
    """
    Returns a Path with the shortest dist from Heathrow to London.

    The allSegs input is a list of Road segment distances. Even though it is a flat
    list of ints, allSegs should be understood in groups of three: A segment, B segment, C segment.
    """
    qpList = toQPList(allSegs)
    # Join the QuadPaths together and then extract the final paths.
    heathrowToLondonPaths = reduce(joinQuadPaths, qpList).paths
    # min takes an iterable, optional additional iterables, optional key for basing
    # comparison, optional default if iterable is empty
    return min(heathrowToLondonPaths, key=Path.figureOfMerit)

def segsToQP(aDist: int, bDist: int=0, cDist: int=0) -> QuadPaths:
    """
    Convert three segment distances into a QuadPaths.
    """
    # for each of the four paths that makes up a QuadPaths, the passed in segment
    # distance is used to create the respective A, B, or C, as a list of the
    # segments needed to get from start to end. The Path is then created by passing
    # in the start, end, and list of steps.
    return QuadPaths([Path(A, A, [A(aDist)]),             # A -> A
                      Path(A, B, [A(aDist), C(cDist)]),   # A -> B
                      Path(B, A, [B(bDist), C(cDist)]),   # B -> A
                      Path(B, B, [B(bDist)])              # B -> B
                      ])

# Allows trace wrapper
@trace
def toQPList(allSegs: List[int]) -> List[QuadPaths]:
    """
    Convert a list of Segment distances into a list of QuadPaths.

    If len(allSegs)%3 != 0, assumes additional segments of length 0.

    Doing it iteratively rather than recursively so that @trace will be called only once.
    """
    qpList: List[QuadPaths] = []
    while len(allSegs) > 0:
        # Uses qpList and allSegs mutably
        # Do you understand how the arguments to segsToQP are passed?

        # segsToQP takes the first three segments in the list of allSegs, unpacks
        # them as arguments using * and appends the resulting QuadPaths to the qpList.
        # allSegs is then updated to be the rest of allSegs after the first three.
        qpList.append(segsToQP(*allSegs[:3]))
        # Do you understand what the following line does?
        allSegs = allSegs[3:]
    return qpList


if __name__ == '__main__':
  # The example from the book.
  dists = [50, 10, 30, 5, 90, 20, 40, 2, 25, 10, 8]
  print(optimalPath(dists))
  # => B->B. Dist: 75. [B/10, C/30, A/5, C/20, B/2, B/8],
